### PR TITLE
Add regression testing infrastructure

### PR DESCRIPTION
**Pull request by brendongo**
Can now specify rules to run on a given repo at a given commit
and will make sure --json output of semgrep does not differ
from some expected output

### CODE REVIEW

**Comment by brendongo** [comment]
Might be more developer friendly if we diffed plain output not --json so we can actually see where it differs.

**Review by ievans — APPROVED** [review]
+1 to diffing plain output, on reflection.

Would be nice to report the time deltas as well but can be done in a future PR

File: semgrep/tests/acceptance/qa.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        print("done clone")
+        subprocess.run(
+            ["git", "checkout", commit_hash],
+            cwd=destination,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+        subprocess.run(
+            ["git", "clean", "-xdf"],
```
Comment (ievans) [inline semgrep/tests/acceptance/qa.py]: @brendongo why is this necessary?

**Comment by brendongo** [comment]
Modified the testing to just check that nothing explodes. So does not actually compare expected and actual output. Just checks to make sure there are no errors.

File: semgrep/tests/acceptance/rule.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+rules:
+  - id: useless-equals
+    pattern: $X==$X
+    message: "Useless equals"
+    languages: [python, javascript]
```
Comment (ievans) [inline semgrep/tests/acceptance/rule.yaml]: You should add all our supported languages here [java, c, go]

Comment (nbrahms) [reply semgrep/tests/acceptance/rule.yaml]: I think the canonical test for NaN in Go might be `$X != $X`.

Comment (brendongo) [reply semgrep/tests/acceptance/rule.yaml]: I think can go in a follow on PR. The given repos are all JS/Python for now.

**Comment by DrewDennison** [comment]
very cool!

File: semgrep/tests/acceptance/qa.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+        yield destination
+
+
+def run_repo(target: str, rewrite: bool = False) -> None:
+    """
+    Runs semgrep on github repo with a simple pattrern
```
Comment (mschwager) [inline semgrep/tests/acceptance/qa.py]: Spell check: `pattrern`

File: semgrep/Makefile [hunk 4]
```diff
@@ -7,3 +7,11 @@ test:
 install:
 	cp -rv ./build/semgrep.dist/ /usr/local/bin/semgrep-files/
 	ln -sf /usr/local/bin/semgrep-files/semgrep /usr/local/bin/semgrep
+
+.PHONY: qa-test
+qa-test: build
+	pytest -s tests/acceptance/qa.py
+
+.PHONY: regenerate-tests
+regenerate-tests:
```
Comment (mschwager) [inline semgrep/Makefile]: Does this do anything right now? I assume we'd have to have a `__main__` check in `qa.py` for this do things.

Comment (brendongo) [reply semgrep/Makefile]: You're right this is stale. Removing.
